numba_dppy/numpy_usm_shared.py

```diff
@@ -0,0 +1,728 @@
+    context.nrt._require_nrt()
+
+    mod = builder.module
+    u32 = ir.IntType(32)
+
+    # Get the Numba external allocator for USM memory.
+    ext_allocator_fnty = ir.FunctionType(cgutils.voidptr_t, [])
+    ext_allocator_fn = mod.get_or_insert_function(
+        ext_allocator_fnty, name="usmarray_get_ext_allocator"
+    )
+    ext_allocator = builder.call(ext_allocator_fn, [])
+    # Get the Numba function to allocate an aligned array with an external allocator.
+    fnty = ir.FunctionType(cgutils.voidptr_t, [cgutils.intp_t, u32, cgutils.voidptr_t])
+    fn = mod.get_or_insert_function(
+        fnty, name="NRT_MemInfo_alloc_safe_aligned_external"
+    )
+    fn.return_value.add_attribute("noalias")
+    if isinstance(align, builtins.int):
+        align = context.get_constant(types.uint32, align)
+    else:
+        assert align.type == u32, "align must be a uint32"
+    return builder.call(fn, [size, align, ext_allocator])
+
+
+_registered = False
+
+
+def is_usm_callback(obj):
+    dprint("is_usm_callback:", obj, type(obj))
+    if isinstance(obj, numba.core.runtime._nrt_python._MemInfo):
+        mobj = obj
+        while isinstance(mobj, numba.core.runtime._nrt_python._MemInfo):
+            ea = mobj.external_allocator
+            dppl_rt_allocator = numba_dppy._dppy_rt.get_external_allocator()
+            dprint("Checking MemInfo:", ea)
+            if ea == dppl_rt_allocator:
+                return True
+            mobj = mobj.parent
+            if isinstance(mobj, ndarray):
+                mobj = mobj.base
+    return False
+
+
+def numba_register():
+    global _registered
+    if not _registered:
+        _registered = True
+        ndarray.add_external_usm_checker(is_usm_callback)
+        numba_register_typing()
+        numba_register_lower_builtin()
+
+
+# Copy a function registered as a lowerer in Numba but change the
+# "np" import in Numba to point to usmarray instead of NumPy.
+def copy_func_for_usmarray(f, usmarray_mod):
+    import copy as cc
+
+    # Make a copy so our change below doesn't affect anything else.
+    gglobals = cc.copy(f.__globals__)
+    # Make the "np"'s in the code use usmarray instead of Numba's default NumPy.
+    gglobals["np"] = usmarray_mod
+    # Create a new function using the original code but the new globals.
+    g = ftype(f.__code__, gglobals, None, f.__defaults__, f.__closure__)
+    # Some other tricks to make sure the function copy works.
+    g = functools.update_wrapper(g, f)
+    g.__kwdefaults__ = f.__kwdefaults__
+    return g
+
+
+def types_replace_array(x):
+    return tuple([z if z != types.Array else UsmSharedArrayType for z in x])
+
+
+def numba_register_lower_builtin():
+    todo = []
+    todo_builtin = []
+    todo_getattr = []
+    todo_array_member_func = []
+
+    # For all Numpy identifiers that have been registered for typing in Numba...
+    # this registry contains functions, getattrs, setattrs, casts and constants...need to do them all? FIX FIX FIX
```

Comment:
```suggestion
    # For all Numpy identifiers that have been registered for typing in Numba...
    # this registry contains functions, getattrs, setattrs, casts and constants...
```
https://github.com/IntelPython/numba-dppy/issues/142